### PR TITLE
Prevent background crash

### LINKED ISSUE #1057: iOS app sometimes crashes in background
The crash happens in the core as we can see in the crash log. 

Nevertheless ChatViewController.startTimer() should be stopped when the app is in background / the display is turned off because the closure in startTimer() is responsible for updating the UI. That's not needed at all in this situation. 

```
Hardware Model:      iPhone13,3
Process:             deltachat-ios [5128]
Path:              
Identifier:          chat.delta
Version:             49 (1.16.0)
AppStoreTools:       12D4d
AppVariant:          1:iPhone13,3:14
Beta:                YES
Code Type:           ARM-64 (Native)
Role:                unknown
Parent Process:      launchd [1]
Coalition:           chat.delta [1283]


Date/Time:           2021-02-13 14:42:13.5452 +0100
Launch Time:         2021-02-13 10:38:01.0939 +0100
OS Version:          iPhone OS 14.4 (18D52)
Release Type:        User
Baseband Version:    1.42.03
Report Version:      104

Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Reason: Namespace RUNNINGBOARD, Code 0xdead10cc
Triggered by Thread:  0

Thread 0 name:
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x00000001be6b4ac8 pread + 8
1   DcCore                        	0x00000001013e4a20 0x100af4000 + 9374240
2   DcCore                        	0x00000001013f275c 0x100af4000 + 9430876
3   DcCore                        	0x00000001013f2460 0x100af4000 + 9430112
4   DcCore                        	0x0000000101416e0c 0x100af4000 + 9580044
5   DcCore                        	0x0000000101411d4c 0x100af4000 + 9559372
6   DcCore                        	0x0000000101418bc8 0x100af4000 + 9587656
7   DcCore                        	0x00000001014035d4 0x100af4000 + 9500116
8   DcCore                        	0x00000001013cf848 0x100af4000 + 9287752
9   DcCore                        	0x0000000101343cb8 0x100af4000 + 8715448
10  DcCore                        	0x00000001010302e0 0x100af4000 + 5489376
11  DcCore                        	0x0000000100bdb374 0x100af4000 + 947060
12  DcCore                        	0x0000000100b8026c 0x100af4000 + 574060
13  DcCore                        	0x0000000100f17840 0x100af4000 + 4339776
14  DcCore                        	0x0000000101095c44 0x100af4000 + 5905476
15  DcCore                        	0x0000000100b008c8 DcContext.getChatMessages(chatId:) + 52 (Wrapper.swift:253)
16  DcCore                        	0x0000000100b008c8 DcContext.getMessageIds(chatId:count:from:) + 72 (Wrapper.swift:31)
17  deltachat-ios                 	0x0000000100477f10 ChatViewController.getMessageIds() + 64 (ChatViewController.swift:746)
18  deltachat-ios                 	0x0000000100477f10 closure #1 in closure #1 in ChatViewController.startTimer() + 124 (ChatViewController.swift:294)
19  deltachat-ios                 	0x0000000100517318 thunk for @escaping @callee_guaranteed () -> () + 20 (<compiler-generated>:0)
20  libdispatch.dylib             	0x000000019035024c _dispatch_call_block_and_release + 32 (init.c:1454)
21  libdispatch.dylib             	0x0000000190351db0 _dispatch_client_callout + 20 (object.m:559)
22  libdispatch.dylib             	0x000000019035f7ac _dispatch_main_queue_callback_4CF + 836 (inline_internal.h:2548)
23  CoreFoundation                	0x00000001906d911c __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16 (CFRunLoop.c:1790)
24  CoreFoundation                	0x00000001906d3120 __CFRunLoopRun + 2508 (CFRunLoop.c:3118)
25  CoreFoundation                	0x00000001906d221c CFRunLoopRunSpecific + 600 (CFRunLoop.c:3242)
26  GraphicsServices              	0x00000001a829e784 GSEventRunModal + 164 (GSEvent.c:2259)
27  UIKitCore                     	0x0000000193112ee8 -[UIApplication _run] + 1072 (UIApplication.m:3253)
28  UIKitCore                     	0x000000019311875c UIApplicationMain + 168 (UIApplication.m:4707)
29  deltachat-ios                 	0x000000010046494c main + 68 (AppDelegate.swift:12)
30  libdyld.dylib                 	0x00000001903926b0 start + 4

```

### PR DESCRIPTION
**Pull request by cyBerta**
fixes #1057 

fixes also a potential memory leak by removing the keyboard observer on viewDidDisappear

### CODE REVIEW

File: deltachat-ios/Chat/ChatViewController.swift [hunk 1]
```diff
@@ -403,7 +420,8 @@ class ChatViewController: UITableViewController {
         super.viewDidDisappear(animated)
         dismissCancelled = false
         AppStateRestorer.shared.resetLastActiveChat()
-        saveDraft()
+        draft.save(context: dcContext)
```
Comment (r10s) [inline deltachat-ios/Chat/ChatViewController.swift:423]: saveDraft() can be removed completely then.

**Review by r10s — COMMENTED** [review]
just a little comment, code-wise, this looks good to me, did not test that out much yet, however.

the timer is for updating the timestamps, right?

**Review by r10s — COMMENTED** [review]
i am wondering if we need `applicationDidBecomeActive` at all, we start the time there - but that is also done at the end of `viewDidAppear` that also should be called.

same for `applicationDidResignActive` and `viewDidDisappear` both call stopTimer() and draft.save().

in similar views, we do not use `applicationDidBecomeActive`/`applicationDidResignActive` - why is it needed here?

if it is needed, however, it lgtm :)

**Comment by cyBerta** [comment]
`applicationDidBecomeActive` and `applicationWillResignActive` handle the app going to the foreground / background.
`viewDidAppear`and `viewDidAppear` handle the situations the view controller gets added to / removed from the view controller stack.
In this particular case we want avoid that the timer runs if the UI is not visible. This includes the situation when the app went to the background as well as the situation where the view controller is not visible because another controller is on top of the stack or the view controller has been completely removed from the stack.

https://github.com/deltachat/deltachat-ios/issues/439 was the reason why we added the observer for the app moving to background. Without saving the draft then, it might get lost if the system kills the app in background. At the same time we also need to save the draft if the app is in foreground and the user navigates back to the chat list. that's why you find `draft.save(context: dcContext)` also in line 423

**Comment by cyBerta** [comment]
@r10s could you check this PR please? Do you have any objections?

**Review by r10s — APPROVED** [review]
lgtm, thanks for clarifying. let's see if that fixes the crash :)

**Pull request merged**